homeassistant/components/media_player/horizon.py

```diff
@@ -0,0 +1,310 @@
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def keys(self):
+        """Return the predefined keys."""
+        return self._keys
+
+    @property
+    def source_list(self):
+        """List of available (german) TV channels (names only)."""
+        return [self._source_list[c] for c in sorted(self._source_list.keys())]
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORT_UM_HORIZON
+
+    @util.Throttle(MIN_TIME_BETWEEN_SCANS, MIN_TIME_BETWEEN_FORCED_SCANS)
+    def update(self):
+        """Update State using the media server running on the Horizon."""
+        if self._client.is_powered_on():
+            self._state = STATE_PLAYING
+        else:
+            self._state = STATE_OFF
+
+    def turn_on(self):
+        """Turn the device on."""
+        if self._state is STATE_OFF:
+            self._send_key(self._keys.POWER)
+
+    def turn_off(self):
+        """Turn the device off."""
+        if self._state is not STATE_OFF:
+            self._send_key(self._keys.POWER)
+
+    def media_previous_track(self):
+        """Channel down."""
+        self._send_key(self._keys.CHAN_DOWN)
+        self._state = STATE_PLAYING
+
+    def media_next_track(self):
+        """Channel up."""
+        self._send_key(self._keys.CHAN_UP)
+        self._state = STATE_PLAYING
+
+    def media_play(self):
+        """Send play command."""
+        self._send_key(self._keys.PAUSE)
+        self._state = STATE_PLAYING
+
+    def media_pause(self):
+        """Send pause command."""
+        self._send_key(self._keys.PAUSE)
+        self._state = STATE_PAUSED
+
+    def media_play_pause(self):
+        """Send play/pause command."""
+        self._send_key(self._keys.PAUSE)
+        if self._state == STATE_PAUSED:
+            self._state = STATE_PLAYING
+        else:
+            self._state = STATE_PAUSED
+
+    def play_media(self, media_type, media_id, **kwargs):
+        """Play media / switch to channel."""
+        if MEDIA_TYPE_CHANNEL == media_type and isinstance(int(media_id), int):
+            self._select_channel(media_id)
+            self._state = STATE_PLAYING
+        else:
+            _LOGGER.error("Invalid type %s or channel %d. Supported type: %s",
+                          media_type, media_id, MEDIA_TYPE_CHANNEL)
+
+    def select_source(self, source):
+        """Select a channel."""
+        if str(source).isdigit():
+            digits = str(source)
+        else:
+            digits = [str(k) for k, v
+                      in self._source_list.items()
+                      if v == source]
+
+        if digits is not None:
+            self._select_channel("".join(digits))
+            self._state = STATE_PLAYING
+
+    def _select_channel(self, channel):
+        """Select a channel (taken from einder library, thx)."""
+        for i in str(channel):
+            key = int(i) + 0xe300
```

Comment:
This looks like too much protocol specific code. It should probably be part of the interface library.